### PR TITLE
[breaking changes] Added global interface

### LINKED ISSUE #384: Need userdive.q Types
# TL;DR

Refs: https://github.com/userdive/agent.js/blob/master/packages/userdive/src/index.ts#L22

likes https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/google.analytics/index.d.ts#L541-L621

### PR DESCRIPTION
**Pull request by 9renpoto**
# TL;DR

Closes #384

### CODE REVIEW

File: packages/userdive/test/index.test.ts [hunk 1]
```diff
@@ -41,4 +41,52 @@ describe('aync loader', () => {
...
+
+  it('should exercise all ga APIs', () => {
+    const _ud = factory()
+    _ud('create', lorem.word(), 'auto')
+    // _ud('create', lorem.word(), {
+    //   cookieName: lorem.word(),
+    //   cookieDomain: internet.domainName(),
+    //   cookieExpires: random.number()
+    // })
+    _ud('create', lorem.word(), 'auto', 'myTracker', {
```
Comment (9renpoto) [inline packages/userdive/test/index.test.ts]: oops, is not support yet.

File: packages/agent/src/agent.ts [hunk 2]
```diff
@@ -22,32 +22,54 @@ export default class Agent {
...
 
-  send (type: HitType, data?: SendData): AgentCore {
-    this.core.send(type, data || location.href)
+  send (type: HitType, data?: FieldsObject | string): AgentCore {
+    if (typeof type !== 'string') {
+      return this.core
+    }
+    switch (type) {
+      case 'pageview':
+        if (!data || typeof data === 'string') {
```
Comment (shoota) [inline packages/agent/src/agent.ts]: Does not need strict undefined check?

Comment (9renpoto) [reply packages/agent/src/agent.ts]: `!data` includes `'' ? 'defined' : "undefined"`

File: docs/events.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+_ud("send", "event", {
+    eventCategory: "Category",
+    eventAction: "Action",
+    eventLabel: "Label",
+    eventValue: 1000
+});
+```
+
+```js
+_ud("send", {
```
Comment (9renpoto) [inline docs/events.md]: WIP, need test

File: package.json [hunk 4]
```diff
@@ -98,14 +98,14 @@
...
-    "test": "lerna run test",
+    "test": "lerna run test --parallel",
     "w:examples": "lerna run dev --scope userdive-examples",
     "w:website": "cd website && npm start"
   },
   "size-limit": [
     {
       "path": "cdn/agent.js",
-      "limit": "8 KB"
+      "limit": "8.2 KB"
```
Comment (9renpoto) [inline package.json]: 😰

File: packages/agent/src/agent.ts [hunk 5]
```diff
@@ -22,37 +22,69 @@ export default class Agent {
...
-  send (type: HitType, data?: SendData): AgentCore {
-    this.core.send(type, data || location.href)
+  send (type: HitType | FieldsObject, data?: FieldsObject | string): AgentCore {
+    if (typeof type === 'object') {
+      data = type
+      type = type.hitType as HitType
+    }
+    let page
+    if (typeof data === 'object') {
+      page = this.set(data).env.l
```
Comment (9renpoto) [inline packages/agent/src/agent.ts]: parsed cached location, before `send.pageview`

File: packages/agent/src/core.ts [hunk 6]
```diff
@@ -112,52 +107,61 @@ export default class AgentCore extends Store {
...
-        ) {
-          this.eventId++
-          this.sendEvent(event)
-        }
-        break
+    const data = getEnv(pathname2href(page))
+    if (!data || !this.baseUrl) {
+      return warning(`failed init`)
+    }
+    this.merge({ type: 'env', data })
```
Comment (9renpoto) [inline packages/agent/src/core.ts]: ここで絶対上書きしてしまうので、引数に絶対新しいものを持ってくる

File: packages/agent/src/core.ts [hunk 7]
```diff
@@ -112,52 +107,61 @@ export default class AgentCore extends Store {
...
+    get(
+      `${this.baseUrl}/${this.loadTime}/env.gif`,
+      obj2query(
+        objectAssign({}, this.get('env'), this.get('custom')) as any /* TODO */
+      ),
+      () => {
+        this.destroy()
+      }
+    )
+    this.set('page', undefined) // remove locale cache
```
Comment (9renpoto) [inline packages/agent/src/core.ts]: `state.env.l` is cached location

update without `send`, like `_ud('set', 'page', internet.url)` `_ud('set', { page: internet.url() })`

File: packages/agent/src/global.ts [hunk 8]
```diff
@@ -1,23 +1,32 @@
...
+  ) => void
+) => (cmd: string, ...args: any[]): void => {
   const [n, command] = cmd.split('.')
   if (n && command) {
     cmd = command
   }
   const trackerName: string = command ? n : 'default'
   /**
-   * _ud('create', 'id', { name: 'myTracker' })
+   * _ud('create', 'id', 'auto', { name: 'myTracker' })
```
Comment (9renpoto) [inline packages/agent/src/global.ts]: not support `_ud('create', 'id', { name: myTracker })`

File: package.json (hunk 4 above)
Comment (shkumagai) [reply package.json]: ﾂﾗｲ…

**Pull request merged**